federated_aws_cli/login.py

```diff
@@ -3,111 +3,176 @@
+        ts = diff.total_seconds()
+
+        if ts < laytime:
+            logger.debug("id_token is about to expire, getting new one")
+            return self.get_wait_for_id_token()
+
+        logger.debug("id_token is still valid: {}".format(self.tokens["expires_at"]))
+
+    def get_wait_for_id_token(self):
+        """
+        Follow the PKCE auth flow by spawning a browser for the user to login,
+        passing a redirect_uri that points to a localhost listener. Once ther user
+        logs into the IdP in the browser, the IdP will redirect the user to the
+        localhost listener, making the OIDC code available to the CLI. CLI then
+        exchanges the code for tokens with the IdP and returns the tokens
+
+        :return: id token dict containing  {'access_token': '...', 'id_token': '...', 'expires_in': 86400, 'expires_at':
+        'date', 'token_type': 'Bearer'}
+        """
+        self.__deferred_init__()
+        url_parameters = {
+            "scope": self.scope,
+            "response_type": "code",
+            "redirect_uri": self.redirect_uri,
+            "client_id": self.client_id,
+            "code_challenge": self.code_challenge,
+            "code_challenge_method": "S256",
+            "state": self.state,
+        }
+        # We don't set audience here because Auth0 will set the audience on it's
+        # own
+        url = "{}?{}".format(self.authorization_endpoint, urlencode(url_parameters))
+
+        # Open the browser window to the login url
+        # Start the listener
+        logger.debug("About to spawn browser window to {}".format(url))
+        webbrowser.get("firefox").open(
+            url
+        )  # This specifies firefox to work around webbrowser.BackgroundBrowser sending stdout/stderr to the console :
+        # https://github.com/python/cpython/blob/783b794a5e6ea3bbbaba45a18b9e03ac322b3bd4/Lib/webbrowser.py#L177-L181
+        logger.debug("About to begin listener on port {}".format(self.port))
+        code, response_state, error_message = listener.get_code(self.port)
+
+        if code is None:
+            logger.error("Something wrong happened, could not retrieve session data")
+            exit(1)
+
+        if self.state != response_state:
+            logger.error("Error: State returned from IdP doesn't match state sent")
+            exit(1)
+
+        if error_message is not None:
+            logger.error(error_message)
+            exit(1)
+
+        # Exchange the code for a token
+        headers = {"Content-Type": "application/json"}
+        body = {
+            "grant_type": "authorization_code",
+            "client_id": self.client_id,
+            "code_verifier": self.code_verifier,
+            "code": code,
+            "redirect_uri": self.redirect_uri,
+        }
+        with requests_cache.disabled():
+            r = requests.post(self.token_endpoint, headers=headers, data=json.dumps(body))
+            data = json.loads(r.text)
```

Comment:
You shouldn't need to use `json.loads`, as `requests` has a `.json()` function you can call on a response.

Reply:
its required for python2

Reply:
It is? I’ve never heard of that before, huh.

Reply:
yeah, for the little story, i always remove it because somehow i tend to forget (must be the old age!), run in p2 and it fails because that method doesnt exist